### PR TITLE
Enhance information about version control (Git)

### PR DESCRIPTION
**Pull request by warpech**
Description in commit messages

Implements https://github.com/Starcounter/Guidelines/issues/13

@Starcounter/sc-apps please review

### CODE REVIEW

File: version_control/pull_request_process.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Contributing Code
+# Pull Request Process
```
Comment (Mackiovello) [inline version_control/pull_request_process.md]: Why did you change the headline? If we're going to change it, I would write something like "Steps for Contributing Code" to make it more descriptive

File: version_control/commit_messages.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# Commit Messages
+
+## Commit messages should say why you made a change, not what you changed.
+
+A good commit message has the following:
+
+- **explanation of the problem**
+- **explanation of the solution**
+- **reference to the relevant GitHub issue** in form `#{number of the issue}`. GitHub recognizes this pattern and shows a cross-reference
```
Comment (Mackiovello) [inline version_control/commit_messages.md]: Does this need to be bold? There's already a lot of bold text in this section, so it loses its effect.

File: version_control/commit_messages.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# Commit Messages
+
+## Commit messages should say why you made a change, not what you changed.
```
Comment (Mackiovello) [inline version_control/commit_messages.md]: Why is this a headline? Can't it be normal text instead? Right now, this text is screaming at me

File: version_control/commit_messages.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+This is a **good commit message**:
+
+```
+Problem: Some elements do not display starting from starcounter-include 3.0.0
+
+This is because they have a slot attribute but there is no respective slot
+element in the declarative Shadow DOM (https://github.com/Starcounter/RebelsLounge/issues/152)
+
+Solution: Add missing slot elements to declarative shadow DOM
+```
```
Comment (Mackiovello) [inline version_control/commit_messages.md]: Is this realistic? I don't think this will be adopted unless it's drilled super hard.

What if I just fixed a typo, do I need to do this? 
```
Problem: There was a typo in the text

Solution: Fix the typo
```

For the example you write, I don't think it's clearer than this:
`Add missing slot elements to declarative shadow DOM, for Starcounter/RebelsLounge#152`. We can assume that the problem is presented in the issue.

File: version_control/commit_messages.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+## Changing the message after committing
+
+If you have not yet pushed your commits, you can still edit the commit messages.
+
+To edit the commit message of the last commit, you can simply change it using `git commit --amend`. Sourcetree and other Git GUIs also give an option to amend a previous commit.
+
+To edit commit messages of any commit you can use Git interactive rebase: https://git-scm.com/book/id/v2/Git-Tools-Rewriting-History
+
+Remember though to never change commits that have been pushed to upstream!
```
Comment (Mackiovello) [inline version_control/commit_messages.md]: Remember though to never change commits that have been pushed to upstream! -> Never change commits that have been pushed

File: version_control/commit_messages.md (hunk 3 above)
Comment (warpech) [inline version_control/commit_messages.md]: For ephasis. If this hurts your eyes, I will change it to normal text.

File: version_control/pull_request_process.md (hunk 1 above)
Comment (warpech) [inline version_control/pull_request_process.md]: For me "Contributing Code" was too generic. This new title says exactly what's inside. Don't you think?

File: version_control/commit_messages.md (hunk 2 above)
Comment (warpech) [inline version_control/commit_messages.md]: Good point.

File: version_control/commit_messages.md (hunk 4 above)
Comment (warpech) [inline version_control/commit_messages.md]: As this [blog post](http://rickardlindberg.me/writing/problem-in-commit-message/) explains, dividing the commit message into `problem / solution` is a mental framework to make sure that you provide both of these parts in your commit message.

`Fix typo` is an obvious fix that indeed is self explanatory. But imagine another case: `rename variable`. Do you think that would be enough for a commit message? I think that not, because I would like to know WHY a variable was renamed. In other words - what was the problem, to which the solution is to rename a variable.

I can rephrase that page to say: 

> The above message is bad because it does not include an explanation of the problem, an explanation of the solution or the reference number. Ideally, it should contain all these parts. 
> 
> A good mental framework is to divide the commit message into two parts: Problem and Solution. See the below example for reference:

Comment (Mackiovello) [inline version_control/commit_messages.md]: > As this blog post explains, dividing the commit message into problem / solution is a mental framework to make sure that you provide both of these parts in your commit message.

Well, as he writes, it's not about the commit message, but rather that you solve an actual problem:
> The reason behind it is that you should only be solving problems, not implement features (that someone might or might not need later).

> In other words - what was the problem, to which the solution is to rename a variable.

I agree, both the problem and the solution should be included.

My suggestion is:
* If a commit fixes an issue, that issue should be referenced
* If there's no relevant issue, the fixed problem should be described in the commit message

For example, if you rename an undescriptive variable, write "rename undescriptive variable" in the commit message. With that, the reader knows exactly what you did and why you did it. If the change is more complex, divide the problem and solution in the commit message:
```
problem: [explanation of problem]

solution: [explanation of solution]
```

File: version_control/pull_request_process.md (hunk 1 above)
Comment (Mackiovello) [inline version_control/pull_request_process.md]: I think "Steps for Contributing Code" explains exactly what it is. The page explains the steps for contributing code to one of our repositories, from cloning to the local machine to merging. In my mind, it's about contributing code - creating a pull request is a part of it, but not all of it.

File: version_control/commit_messages.md (hunk 3 above)
Comment (Mackiovello) [inline version_control/commit_messages.md]: Having it as the first thing on the page should be enough emphasis.

File: version_control/commit_messages.md (hunk 2 above)
Comment (joozek78) [inline version_control/commit_messages.md]: If you refer to an issue from another repo, use `org/repo#issue`

File: version_control/commit_messages.md (hunk 4 above)
Comment (joozek78) [inline version_control/commit_messages.md]: BTW it's not a good commit message, it's missing the headline (problem description is not a good headline, it doesn't describe what has been changed)

Comment (warpech) [inline version_control/commit_messages.md]: There is the headline, separated with a single empty line from the body. It just happens that the headline states the problem, not the solution. Anything seriously wrong with that?

File: version_control/commit_messages.md (hunk 2 above)
Comment (warpech) [inline version_control/commit_messages.md]: Thanks, I will add that

File: version_control/commit_messages.md (hunk 4 above)
Comment (warpech) [inline version_control/commit_messages.md]: > BTW it's not a good commit message, it's missing the headline (problem description is not a good headline, it doesn't describe what has been changed).

There is the headline, separated with a single empty line from the body. It just happens that the headline states the problem, not the solution. Anything seriously wrong with that?

> For example, if you rename an undescriptive variable, write "rename undescriptive variable" in the commit message. With that, the reader knows exactly what you did and why you did it. If the change is more complex, divide the problem and solution in the commit message:

Totally agree. I am not arguing that all commit messages should be written in the `Problem / Solution` form. I mean it is a mental picture for writing commit messages that can be also used as a literal pattern wherever helpful

Comment (joozek78) [inline version_control/commit_messages.md]: > There is the headline, separated with a single empty line from the body. It just happens that the headline states the problem, not the solution. Anything seriously wrong with that?

I think that the headline should quickly describe the change and problem doesn't always do that. I'm not sure it's a good idea to discuss it here, but I think we (SC) currently use the headline that way. If you want to convince everybody to state a problem there, I will gladly discuss it, but maybe move it to slack / have a separate issue for it?

File: version_control/pull_request_process.md (hunk 1 above)
Comment (warpech) [inline version_control/pull_request_process.md]: Fair enough.

File: version_control/commit_messages.md (hunk 4 above)
Comment (warpech) [inline version_control/commit_messages.md]: Added some clarification that the problem/solution guideline is voluntary. I hope this is acceptable now. Time will show if it is useful. I will coach using it because I find big value in it. 

If you are in doubt, I encourage everyone to look at your own random commit message older than few months. Can you tell why that change was made? If it is a bug fix, the GitHub issue will usually tell only what was the symptom, not what was the actual problem.

**Comment by warpech** [comment]
Thanks for the review. I think it can be merged
